metaflow/runner/utils.py

```diff
@@ -74,6 +74,32 @@ def read_from_file_when_ready(
         return content
 
 
+async def async_read_from_file_when_ready(
+    file_path: str, command_obj: "CommandManager", timeout: float = 5
+):
+    await asyncio.wait_for(command_obj.process.wait(), timeout)
```

Comment:
Here, we wait for the process to finish with a timeout..

I am not sure we want that. The file content can be written earlier than the finish time of the process.

And we probably want the file content back as soon as possible..

The synchronous version doesn't wait for the process to finish before checking the file content. It keeps reading the file in a loop, and as soon as there is content, it returns it—even if the process is still running.

Reply:
Didn't realize the `poll` call isn't blocking. I rewrote it so that its the same as the sync version except a call to `asyncio.sleep`.

The impl doesn't really sit right with me though, I can imagine a scenario in which a call to `read` doesn't return the entire file content, because another write is issued by the subprocess immediately after. I think we could reimplement this with pipes and poll to see if the write side (subprocess) has closed the pipe and thus indicated it finished the writing. I could take a look at this after.

Reply:
So, we for know for a fact that the file being written will only be written once..

This makes things a bit easier..

Reply:
And we want to get the content ASAP since that allows us to fetch the `Run` object..
(while the process is still running..)

Reply:
@filipcacky has a point here though -- it is true we write once and in *most* cases it should work fine but it is up to the OS to determine how the writes proceed and there is no guarantee that the content will be written all at once. Another mechanism may be warranted. File locks are always a bit annoying though (and not always reliable). We could also check for the presence of a particular string at the end or something (the OS should ensure writes are at least done sequentially). It may be worth fixing now because these types of bugs are particularly nasty when they do occur (it's rare but not impossible).

Reply:
It's probably not worth doing in this PR, but when I have a bit more time I can take a look at it).

Reimplementing it for a peace of mind and or better ergonomics on the write side (not being forced to make only a single write) and faster read on the read side (reading until a pipe close, instead of waking up and checking for a non-null content) is I think worth it.

Reply:
I guess we can resolve this comment for now? But if possible..
@filipcacky can you take a stab at it within this PR itself?

Reply:
Alright, reimplemented it via a FIFO, made the same changes to Deployer.